lib/wasm/src/environ/spec.rs

```diff
@@ -245,11 +245,16 @@ pub trait ModuleEnvironment<'data> {
     /// Get the information needed to produce Cranelift IR for the current target.
     fn target_config(&self) -> TargetFrontendConfig;
 
+    /// Provides the number of signatures up front. By default this does nothing, but
+    /// implementations can use this to preallocate memory if desired.
+    fn reserve_signatures(&mut self, _num: u32) {}
```

Comment:
I'm generally in favor of matching the spec's names. In this case, the thing it's passing in is an `ir::Signature`, which is related to a wasm function type, but also contains a calling convention and potentially `AbiParam` information.

One option would be to also rename `ir::Signature` to `ir::FuncType` too.

Another would be to adjust the API so that we pass the `wasmparser::FuncType` straight from wasmparser into `declare_func_type` instead of translating it into an `ir::Signature` first. Maybe we should also change `declare_global_import`/`declare_table_import`/etc. -- everything in sections_translator.rs that calls `type_to_type` -- to pass through `wasmparser` types, rather than Cranelift IR types.

I don't currently have a clear view of how this API should best evolve. Do you have thoughts here?